docs/05-authentication-and-authorization.md

```diff
@@ -1,14 +1,255 @@
+
+```html
+<UserStateProvider>
+    <Router AppAssembly=typeof(Program).Assembly />
+</UserStateProvider>
+```
+
+At first this will appear to do nothing, but in fact this has made available a *cascading parameter* to all descendant components. A cascading parameter is a parameter that isn't passed down just one level in the hierarchy, but through any number of levels.
+
+## Displaying login state
+
+Create a new component called `UserInfo` in the client project's `Shared` folder, containing:
+
+```html
+Logged in: @UserState.IsLoggedIn
+
+@functions {
+    [CascadingParameter] UserStateProvider UserState { get; set; }
+}
+```
+
+This is similar to using `[Parameter]`, except that it receives cascading values from any ancestor component.
+
+By default, the framework matches by type. We're asking for an instance of `UserStateProvider`, so the framework will give us the closest one in the ancestry. Or if there isn't one, we'll get `null`. It's also possible to match by a string-valued name, but we don't need that here.
+
+It might seem surprising that the `UserStateProvider` supplies *itself* to descendants. In fact, that's a commonly useful pattern, because it makes it easy to expose `public` methods that descendants can invoke. In this case, we'll be calling its `SignIn` and `SignOut` methods soon. However, there's no requirement to follow this pattern: components can supply other values beside themselves as cascading values if they want.
+
+We don't really just want to display `Logged in: false`, so replace the markup in `UserState` as follows:
+
+```html
+<div class="user-info">
+    @if (UserState.CurrentUser == null)
+    {
+        <text>...</text>
+    }
+    else if (UserState.CurrentUser.IsLoggedIn)
+    {
+        <img src="img/user.svg" />
+        <div>
+            <span class="username">@UserState.CurrentUser.DisplayName</span>
+            <span class="sign-out" onclick="@UserState.SignOut">Sign out</span>
+        </div>
+    }
+    else
+    {
+        <button onclick="@UserState.SignIn" class="sign-in">Sign in</button>
+    }
+</div>
+```
+
+This handles three scenarios:
+
+1. The client is waiting for the server to say whether the user is signed in or not
+2. The server says they are signed in
+3. The server says they are not signed in
+
+Finally, let's put the `UserState` in the UI somewhere. Open `MainLayout`, and update the `<div class="top-bar">` as follows:
+
+```html
+<div class="top-bar">
+    (... leave existing content in place ...)
+
+    <UserInfo />
+</div>
+```
+
+Because the user isn't yet signed in, it will display a "sign in" button.
+
+![image](https://user-images.githubusercontent.com/1101362/51807139-0a431100-227b-11e9-923b-b9313abf8992.png)
+
+## Signing in with Twitter
+
+This application uses server-based authentication. The mechanism is as follows:
+
+1. The client asks the server whether the user is logged in.
```

Comment:
Are there any request forgery concerns with this setup? I'm not saying there are - I just want to make sure we've thought about this, since we are using cookies.

Reply:
Same-site cookies are a basic defense. We can add traditional anti-forgery if we want.